sql/src/test/java/io/crate/integrationtests/disruption/seqno/ConcurrentSeqNoVersioningIT.java

```diff
@@ -0,0 +1,756 @@
+        }
+
+        Version previousSeqNo(int decrement) {
+            return new Version(primaryTerm, Math.max(seqNo - decrement, 0));
+        }
+
+        @Override
+        public String getWriteableName() {
+            return "version";
+        }
+
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeLong(primaryTerm);
+            out.writeLong(seqNo);
+        }
+
+        Version previousTerm() {
+            return new Version(primaryTerm - 1, seqNo);
+        }
+
+        Version nextTerm() {
+            return new Version(primaryTerm + 1, seqNo);
+        }
+    }
+
+    private static class AtomicVersion {
+        private final AtomicReference<Version> current;
+
+        private AtomicVersion(Version initialVersion) {
+            this.current = new AtomicReference<>(initialVersion);
+        }
+
+        public Version get() {
+            return current.get();
+        }
+
+        public void consume(Version version) {
+            if (version == null)
+                return;
+            this.current.updateAndGet(current -> version.compareTo(current) <= 0 ? current : version);
+        }
+    }
+
+    private class Partition {
+        private final String id;
+        private final AtomicVersion latestSuccessfulVersion;
+        private final AtomicVersion latestObservedVersion;
+        private final Version initialVersion;
+        private final LinearizabilityChecker.History history = new LinearizabilityChecker.History();
+
+        private Partition(String id, Version initialVersion) {
+            this.id = id;
+            this.latestSuccessfulVersion = new AtomicVersion(initialVersion);
+            this.latestObservedVersion = new AtomicVersion(initialVersion);
+            this.initialVersion = initialVersion;
+        }
+
+        // latest version that was observed, possibly dirty read of a write that does not survive
+        Version latestObservedVersion() {
+            return latestObservedVersion.get();
+        }
+
+        // latest version for which we got a successful response on a write.
+        Version latestSuccessfulVersion() {
+            return latestSuccessfulVersion.get();
+        }
+
+        Consumer<HistoryOutput> invoke(Version version) {
+            int eventId = history.invoke(version);
+            logger.debug("invocation partition ({}) event ({}) version ({})", id, eventId, version);
+            return output -> consumeOutput(output, eventId);
+        }
+
+        private void consumeOutput(HistoryOutput output, int eventId) {
+            history.respond(eventId, output);
+            logger.debug("response partition ({}) event ({}) output ({})", id, eventId, output);
+            latestObservedVersion.consume(output.getVersion());
+            if (output instanceof IndexResponseHistoryOutput) {
+                latestSuccessfulVersion.consume(output.getVersion());
+            }
+        }
+
+        boolean isLinearizable() {
```

Comment:
looks like this is only used in `assertLinearizable` so we could merge both methods (actually the original code at ES does only have the `assertLinearizable` method.